lib/bap_types/bap_ir.ml

```diff
@@ -143,18 +143,67 @@ type sub = {
 
 
 type path = int array
-  [@@deriving bin_io, compare, sexp]
-
-
-type program = {
-  subs  : sub term array;
-  paths : path Tid.Table.t;
-} [@@deriving bin_io, fields, sexp]
-
-let compare_program x y =
-  let compare x y = [%compare:sub term array] x y in
-  compare x.subs y.subs
+[@@deriving bin_io, compare, sexp]
+
+module Program : sig
+  type t = private {
+    subs  : sub term array;
+    paths : path Tid.Table.t;
+  } [@@deriving bin_io, compare, fields, sexp]
+
+  val empty : unit -> t
+
+  val update : t -> sub term array -> t
+
+end = struct
+  type t = {
+    subs  : sub term array;
+    paths : path Tid.Table.t;
+  } [@@deriving bin_io, fields, sexp]
+
+  let mangle_name addr tid name =
+    match addr with
+    | Some a ->
+      sprintf "%s@%s" name @@
+      Bap_bitvector.string_of_value ~hex:true a
+    | None -> sprintf "%s%%%s" name (Tid.to_string tid)
+
+  let mangle_sub s =
+    let addr = Dict.find s.dict Bap_attributes.address in
+    let name = mangle_name addr s.tid s.self.name in
+    Tid.set_name s.tid name;
+    let self = {s.self with name} in
+    {s with self}
+
+  let fix_names olds news =
+    let is_new tid name =
+      match Array.find olds ~f:(fun s -> Tid.equal s.tid tid) with
+      | Some s -> String.(name <> s.self.name)
+      | None -> true in
+    let keep_name tids name tid = Map.add tids ~key:name ~data:tid in
+    let tids = Array.fold news ~init:String.Map.empty ~f:(fun tids sub ->
+        match Map.find tids sub.self.name with
+        | None -> keep_name tids sub.self.name sub.tid
+        | Some _ ->
+          if is_new sub.tid sub.self.name then
+            keep_name tids sub.self.name sub.tid
+          else tids) in
+    Array.map news ~f:(fun sub ->
```

Comment:
`Array.map` will create a new array and copy the contents of the old array even if the `f` function is `ident`:
```
let xs = Array.init 8 ~f:ident;;
val xs : int Core_kernel.Array.t = [|0; 1; 2; 3; 4; 5; 6; 7|]
# let ys = Array.map xs ~f:ident;;
val ys : int Core_kernel.Array.t = [|0; 1; 2; 3; 4; 5; 6; 7|]
# phys_equal xs ys;;
- : bool = false
```

In our case, it will result in double copying of the whole program even if the change was benign and didn't break the naming injectivity. 

So you need to prefix the map function with the check `if Array.length news <> Map.length tids then Array.map ... else news`